docs/src/app/docs/customization/page.mdx

```diff
@@ -126,3 +126,36 @@ export const env = createEnv({
   extends: [authEnv],
 });
 ```
+
+## Further refinement or transformation
+
+You can use the `createFinalSchema` option to further refine or transform the environment variables.
+
+```ts title="src/env.ts"
+import { createEnv } from "@t3-oss/env-core";
+import { z } from "zod";
+
+export const env = createEnv({
+  server: {
+    SKIP_AUTH: z.boolean().optional(),
+    EMAIL: z.string().email().optional(),
+    PASSWORD: z.string().min(1).optional(),
+  },
+  // ...
+  createFinalSchema: (shape, isServer) =>
+    z.object(shape).transform((env, ctx) => {
+      if (env.SKIP_AUTH || !isServer) return { SKIP_AUTH: true } as const;
+      if (!env.EMAIL || !env.PASSWORD) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: "EMAIL and PASSWORD are required if SKIP_AUTH is false",
+        });
+        return z.NEVER;
+      }
+      return {
+        EMAIL: env.EMAIL,
+        PASSWORD: env.PASSWORD,
+      };
+    }),
```

Comment:
i have mixed feelings about this api - on one hand it's great as you can do "post-validation" across all env types (client, shared, server) but it also feels a bit magic.

Reply:
open to suggestions :) it just seems like the simplest way to me to enable further refinement while still passing the raw shape

Reply:
it also means we can pass off correct inference and checking of optional fields etc back to the schema library - currently our logic is basically "if the result is possibly undefined, make it optional as well"

Reply:
yea i think it's fine